### PR TITLE
add insert!

### LINKED ISSUE #74: add insert!
It should be easy to add `insert!` support for PooledVector.

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/PooledArrays.jl/issues/74

### CODE REVIEW

File: src/PooledArrays.jl [hunk 1]
```diff
@@ -590,11 +590,16 @@ end
...
 ##############################################################################
 
-function Base.push!(pv::PooledVector{S,R}, v::T) where {S,R,T}
+function Base.push!(pv::PooledVector, v)
     push!(pv.refs, getpoolidx(pv, v))
     return pv
 end
 
+function Base.insert!(pv::PooledVector, i::Integer, v)
+    insert!(pv.refs, i, getpoolidx(pv, v))
```
Comment (bkamins) [inline src/PooledArrays.jl:599]:
```suggestion
    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
    1 <= i <= length(pv.refs) + 1 || throw(BoundsError("attempt to insert to a vector with length $(length(pv)) at index $i"))
    insert!(pv.refs, i, getpoolidx(pv, v))
```

**Review by quinnj — APPROVED** [review]
LGTM; we don't need to worry about the copy-on-write business at all here or in `push!`, right?

**Comment by bkamins** [comment]
Copy-on-write is handled here https://github.com/JuliaData/PooledArrays.jl/blob/main/src/PooledArrays.jl#L567, which gets called if a new entry needs to be added.

However `push!` and `insert!` are not thread safe - should me mention this in the docs or this is obvious?

**Comment by quinnj** [comment]
Ah, that's right; I missed that it gets handled from the `getpoolidx` call. Yeah, I think it'd be worth mentioning that they aren't threadsafe; it would hopefully be obvious to most users, since they aren't threadsafe for Base `Array`, but it doesn't hurt as a reminder.

File: src/PooledArrays.jl [hunk 2]
```diff
@@ -590,11 +590,20 @@ end
 ##
 ##############################################################################
 
-function Base.push!(pv::PooledVector{S,R}, v::T) where {S,R,T}
+function Base.push!(pv::PooledVector, v)
```
Comment (bkamins) [inline src/PooledArrays.jl]:
```suggestion
function Base.push!(pv::PooledVector, v) # this function is not thread safe
```

File: src/PooledArrays.jl [hunk 3]
```diff
@@ -590,11 +590,20 @@ end
...
     push!(pv.refs, getpoolidx(pv, v))
     return pv
 end
 
+function Base.insert!(pv::PooledVector, i::Integer, v)
```
Comment (bkamins) [inline src/PooledArrays.jl]:
```suggestion
function Base.insert!(pv::PooledVector, i::Integer, v) # this function is not thread safe
```

**Comment by bkamins** [comment]
OK - I have added comments (I think it does not make much sense to add a docstring as it would be the same as for general `push!` and `insert!`)

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

I will make a release.

**Pull request merged**